content/blog/2019-10-25-efficient-path-profiling/index.md

```diff
@@ -0,0 +1,109 @@
++++
+title = "Efficient Path Profiling "
+extra.author = "Dietrich Geisler"
+extra.bio = """
+  [Dietrich Geisler](https://www.cs.cornell.edu/~dgeisler/) is a 3rd year PhD student researching Language Design and Compilers.  Enjoys gaming and climbing.
+"""
++++
+
+## Introduction
+
+Identifying path bottlenecks on program execution is critical when evaluating and optimizing program performance.  
+Due to the exponential number of paths in large programs, however, counting the most frequented path in a given execution directly is intractable for any realistic program.
+Solutions to this problem long focused on counting other execution properties dynamically, such as the number of times each control flow graph (CFG) edge is taken.
+
+Such an edge-focused algorithm, however, does not allow unique construction of paths, and cannot accurately predict the path profile of a given execution.
+The usual definition for the most taken path is defined greedily, where it is constructed as the series of maximally taken paths for each node.
+Consider, for instance, the edge profiles, which _both_ compose to produce the path profile shown:
+
+<img src="bad_paths.png" alt="drawing" width="400"/>
+
+Neither measurements, however, predict the path _ABCDEF_ as the most taken, despite it being defined as such by the usual heuristic.
+Indeed, it was long thought that this loss of accuracy was necessary when defining a sufficiently fast algorithm.
+
+We will explore how Thomas Ball's and James R. Larus' paper [_Efficient Path Profiling_](https://dl.acm.org/citation.cfm?id=243857) define an efficient algorithm for dynamically computing information about path use.
+This path profiling algorithm computes information about paths directly, and so does not suffer from the same loss of accuracy as approximations derived from edge use measurements.
+As promised by the title, this algorithm is both memory efficient and fast enough to use when profiling realistic programs.
+Experimentally, path profiling is shown to only have a 31% profiling overhead compared with edge profiling having a 16% overhead on a standard benchmark suite.
+
+## The Algorithm for DAGs
+
+We start by assuming profiling over a directed acyclic graph (DAG), but will later explore extending to loops.
+Any DAG will have at least node with no incoming edges, called entry nodes, and at least node with no outgoing edges, called exit nodes.
+We extend any CFG with a single entry and single exit node, thus making each unique.
+The steps of the path profiling algorithm are as follows:
+1. Assign a minimal number of integer values to edges such that any calculated edge sum is unique.
```

Comment:
This sentence could clarify a tiny bit what you mean by "calculated edge sum": namely, the idea is that you assign integers to edges such that, when you sum the edge values along any path, that sum is unique among all paths.